redis/src/cluster_async/connections_logic.rs

```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
+            // Both connections were successfully established
+            let (mut user_conn, mut user_ip): (C, Option<IpAddr>) = conn_1;
+            let (mut management_conn, management_ip): (C, Option<IpAddr>) = conn_2;
+            if user_ip == management_ip {
+                // Set up both connections
+                if let Err(err) = setup_user_connection(&mut user_conn, params).await {
+                    return err.into();
+                }
+                match setup_management_connection(&mut management_conn).await {
+                    Ok(_) => ConnectAndCheckResult::Success(create_async_node(
+                        user_conn,
+                        Some(management_conn),
+                        user_ip,
+                    )),
+                    Err(err) => ConnectAndCheckResult::ManagementConnectionFailed {
+                        node: create_async_node(user_conn, None, user_ip),
+                        err,
+                    },
+                }
+            } else {
+                // Use only the connection with the latest IP address
+                warn_mismatch_ip(addr, user_ip, management_ip);
+                if has_dns_changed(addr, &user_ip.unwrap()).await {
+                    // The user_ip is incorrect. Use the created `management_conn` for the user connection
+                    user_conn = management_conn;
+                    user_ip = management_ip;
+                }
+                match setup_user_connection(&mut user_conn, params).await {
+                    Ok(_) => ConnectAndCheckResult::ManagementConnectionFailed {
+                        node: create_async_node(user_conn, None, user_ip),
+                        err: (ErrorKind::IoError, "mismatched IP").into(),
+                    },
+                    Err(err) => err.into(),
+                }
+            }
+        }
+        (Ok(conn), Err(err)) | (Err(err), Ok(conn)) => {
+            // Only a single connection was successfully established. Use it for the user connection
+            warn_management_conn_faild(addr, &err);
+            let (mut user_conn, user_ip): (C, Option<IpAddr>) = conn;
+            match setup_user_connection(&mut user_conn, params).await {
+                Ok(_) => ConnectAndCheckResult::ManagementConnectionFailed {
+                    node: create_async_node(user_conn, None, user_ip),
+                    err,
+                },
+                Err(err) => err.into(),
+            }
+        }
+        (Err(err_1), Err(err_2)) => {
+            // Neither of the connections succeeded.
+            RedisError::from((
+                ErrorKind::IoError,
+                "Failed to refresh both connections",
+                format!(
+                    "Node: {:?} received errors: `{:?}`, `{:?}`",
+                    addr, err_1, err_2
+                ),
+            ))
+            .into()
+        }
+    }
+}
+
+async fn create_management_connection<C>(
```

Comment:
I think this function is redundant, we can add the warning into the ManagementConnectionFailed result constructor